---
title: CA2124:在外部 try 块中包装易受攻击的 finally 子句
description: 在版本 1.0 和 1.1 的 .NET Framework 中，公共或受保护的方法包含 try/catch/finally 块。 finally 块似乎要重置安全状态，并且不包括在某个 finally 块中。
ms.date: 11/04/2016
ms.topic: reference
f1_keywords:
- CA2124
- WrapVulnerableFinallyClausesInOuterTry
helpviewer_keywords:
- CA2124
- WrapVulnerableFinallyClausesInOuterTry
ms.assetid: 82efd224-9e60-4b88-a0f5-dfabcc49a254
author: mikejo5000
ms.author: mikejo
manager: jmartens
ms.technology: vs-ide-code-analysis
ms.workload:
- multiple
ms.openlocfilehash: 54e57bbc6440d5d63e94e69d108831dfdcedb438
ms.sourcegitcommit: b12a38744db371d2894769ecf305585f9577792f
ms.translationtype: HT
ms.contentlocale: zh-CN
ms.lasthandoff: 09/13/2021
ms.locfileid: "126601428"
---
# <a name="ca2124-wrap-vulnerable-finally-clauses-in-outer-try"></a>CA2124:在外部 try 块中包装易受攻击的 finally 子句

|项目|“值”|
|-|-|
|RuleId|CA2124|
|Category|Microsoft.Security|
|重大更改|非中断|

## <a name="cause"></a>原因
在版本 1.0 和 1.1 的 .NET Framework 中，公共或受保护的方法包含 `try`/`catch`/`finally` 块。 `finally` 块似乎要重置安全状态，并且不包括在某个 `finally` 块中。

> [!NOTE]
> 此规则已被弃用。 有关详细信息，请参阅[已弃用的规则](fxcop-unported-deprecated-rules.md)。

## <a name="rule-description"></a>规则说明
此规则会在面向版本 1.0 和 1.1 的 .NET Framework 的代码中查找可能容易受到调用堆栈中恶意异常筛选器攻击的 `try`/`finally` 块。 如果 try 块中发生模拟等敏感操作并引发异常，筛选器可以在 `finally` 块之前执行。 对于模拟示例，这意味着筛选器将作为模拟用户执行。 筛选器目前只能在 Visual Basic 中实现。

> [!NOTE]
> 在 2.0 及更高版本的 .NET Framework 中，如果重置直接在包含异常块的方法中发生，运行时会自动保护 `try`/`catch`/ `finally` 块免受恶意异常筛选器攻击。

## <a name="how-to-fix-violations"></a>如何解决冲突
将展开的 `try`/`finally` 放置在外部 try 块中。 请参阅以下第二个示例。 这会强制 `finally` 在筛选器代码之前执行。

## <a name="when-to-suppress-warnings"></a>何时禁止显示警告
不禁止显示此规则发出的警告。

## <a name="pseudo-code-example"></a>伪代码示例

### <a name="description"></a>说明

下面伪代码说明此规则检测到的模式。

```csharp
try {
   // Do some work.
   Impersonator imp = new Impersonator("John Doe");
   imp.AddToCreditCardBalance(100);
}
finally {
   // Reset security state.
   imp.Revert();
}
```

以下伪代码展示了可以用于保护代码并满足此规则的模式。

```csharp
try {
     try {
        // Do some work.
     }
     finally {
        // Reset security state.
     }
}
catch()
{
    throw;
}
```
